### ISSUE TITLE
pam_access.so considers tty* names as hostnames

### ISSUE THREAD

**Issue by rpigott** [linux-pam/linux-pam#834]
Yes I have seen #345 and #711, and I am using pam 1.6.1.

```
$ pacman -Q pam openssh
pam 1.6.1-3
openssh 9.8p1-2
$ cat /etc/pam.d/sshd
#%PAM-1.0

auth      include   system-remote-login
account   required  pam_access.so
account   include   system-remote-login
password  include   system-remote-login
session   include   system-remote-login
```

The sample access.conf shipped with pam contains the following

https://github.com/linux-pam/linux-pam/blob/63ba6e4aa17761461fbddf71af7cb154f1e5b9ae/modules/pam_access/access.conf#L81-L82

If I uncomment this configuration, I can see that pam_access considers every token a hostname.

<details><summary>trace of getaddrinfo calls in sshd during login</summary>

```
$ cat gai.bt
#!/bin/bpftrace

uprobe:/usr/lib/libc.so.6:getaddrinfo /arg0 != 0/
{
	@start[tid] = nsecs;
	@host[tid] = str(arg0);
}
uretprobe:/usr/lib/libc.so.6:getaddrinfo /@start[tid] > 0/
{
	$diff = nsecs - @start[tid];
	printf("%s %s %d ms\n", comm, @host[tid], $diff/1e6);
}
END
{
	clear(@start);
	clear(@host);
}
$ grep -v '^#' /etc/security/access.conf
+:rpigott:cron crond :0 tty1 tty2 tty3 tty4 tty5 tty6
```
```
$ sudo bpftrace gai.bt # login as rpigott from my laptop
Attaching 3 probes...
sshd-session cron 2 ms
sshd-session crond 1 ms
sshd-session :0 1 ms
sshd-session tty1 1 ms
sshd-session tty2 1 ms
sshd-session tty3 0 ms
sshd-session tty4 0 ms
sshd-session tty5 0 ms
sshd-session tty6 0 ms
sshd-session cron 0 ms
sshd-session crond 0 ms
sshd-session :0 0 ms
sshd-session tty1 0 ms
sshd-session tty2 0 ms
sshd-session tty3 0 ms
sshd-session tty4 0 ms
sshd-session tty5 0 ms
sshd-session tty6 0 ms
^C
```
</details>

So we see dns lookups for cron, crond, :0, and the tty names. To me this is not expected behavior — the documentation makes it sound as if at least the tty* values are intended only to match tty names, and my intention was not for any of these to be considered hostnames but _only_ service/tty names.

I really think there should probably be a configuration that only permits qualified hostnames, and tokens should then be consider hostnames only if they contain a dot, otherwise I will accidentally permit hosts named "tty*".

**Comment by ldv-alt** [comment]
Is this a suggestion to disqualify host names like `localhost`, despite of the rationale given in #447?

**Comment by rpigott** [comment]
No, it would just be spelled `localhost.` instead.

I don't care for the hostname feature really, I'm just trying to offer a solution that hopefully satisfies those who do rather than just delete it. I do think the current behavior is clearly bad, though, and needs to be changed. On my lan, and I suspect many more where dns is tied to dhcp hostnames, `hostnamectl hostname tty1 && networkctl renew wlan0` is enough to by pass a basic config when authenticating from a remote host.

**Comment by ldv-alt** [comment]
@thkukuk, what do you think about it?

**Comment by rpigott** [comment]
@thkukuk ping?

**Comment by thkukuk** [comment]
> @thkukuk ping?

Sometimes people are on FTO...

From bug reports I know that many people use the host name feature for pam_access, so I'm reluctant to change that behavior, as this could break many setups and would bring up the next wave of complains because of incompatible changes. So currently, I know much more bug reports about the hostname feature than complains (I'm only aware of this one).
And your hostnamectl example can only be used by root, not by a normal user, and root can do everything. So not really relevant.

And yes, pam_access is pretty old and enhanced so often, that it needs a complete, redesigned successor, which then could be incompatible.

I would not touch this.

**Comment by rpigott** [comment]
> And your hostnamectl example can only be used by root, not by a normal user, and root can do everything. So not really relevant.

Hm? These commands are used to demonstrate changing the _remote_ host name is sufficient to bypass pam_access on the _target_ host. No privilege on the target host is required. I really think the current behavior is bad.

> this could break many setups and would bring up the next wave of complains because of incompatible changes.

Setups already were broken though. The hostname feature introduced an incompatible change when it modified tokens like "tty" or "crond" to match hostnames. Whether users notice it or not, many previously effective pam_access configurations are no longer effective for their intended purpose.

**Comment by rpigott** [comment]
@ldv-alt Do you have an opinion?

**Comment by carnil** [comment]
I guess this got a CVE assigned: https://www.cve.org/CVERecord?id=CVE-2024-10963 The only reference is https://bugzilla.redhat.com/show_bug.cgi?id=2324291

**Comment by rpigott** [comment]
I emailed RH with a link to this issue last week.

I did not request a CVE, but tbh I would have been surprised if one was not assigned. The behavior is obviously a problem imo.

**Comment by thkukuk** [comment]
The problem is that you have an untrusted DNS setup but are using access control. This is not common standard in "enterprise" networks, maybe home networks. This setup would not even pass a security audit. That's why most people don't see this as real problem.

But there are several real problems with the current implementation, one is a huge delay in trying to resolve TTY names. I made another PR, which does not try to resolve TTY names and where you can disable resolving of tokens as hostnames.
Only open is to detect that a token is a PAM service name and don't resolve that, too. But that's difficult, as there is no way to get a list of this names.

**Comment by pemensik** [comment]
DNSSEC Is completely unrelated. Take please into account that single label names are resolved by LLMNR protocol by default. Even unprivileged remote user can monitor queries on port 5355 and send bogus replies. Mixing pty names is just a terrible idea. Could be hostnames required to be prefixed by ``DNS:`` for example? IP address has recognizable syntax, but hostname is much less limited. Because it would be a CVE fix, I thing possible backward incompatibility is acceptable.

**Comment by ldv-alt** [comment]
Changing `access.conf` syntax to require hostnames to be prefixed or syntactically marked in any other way would be essentially the same as disabling hostnames support as those applying a CVE fix are _not_ expected to review and update their deployed `access.conf` files.

Let's admit this fact: there is no way to tell service names from hostnames, so fixing this CVE would require disabling support of hostnames, either directly, or indirectly by changing `access.conf` syntax, e.g. by requiring hostnames to be FQDN.

**Comment by rpigott** [comment]
We already have a module that can make explicit comparisons in pam_succeed_if. I don't think it's that valuable to introduce a bunch of new syntax to access.conf. @thkukuk is right to point out that pam_access is old — I expect approximately nobody to go change their access.conf configuration.

I would also like to avoid backwards incompatible changes. IIUC pam_access is older than me and it had no hostname support for almost that entire time. The hostname feature was a backwards incompatible change, that surreptitiously transformed the meaning of existing configurations into allowing remote hosts. It did so without a mention in the NEWS file, and was then widely backported in distros like RHEL, thinking it a bugfix, I assume, which indicates to me that downstream did not understand this change. In my experience, it is vanishingly rare we can trust getaddrinfo to be accurate and tamper-proof, so I would not have opted-in to this change, given the choice, but until yesterday there was not a way to disable it. 

From that perspective, removing the hostname feature is hardly a compatibility issue in my view. The relatively few users who were aware of this change and modified their configurations to make use of it can do so again if required, either with the FQDN patch or some new syntax with an RHOST= prefix or something. FQDN was the most "compatible" implementation of this feature that I could come up with, but if another syntax is more palatable please go with that.

**Comment by thkukuk** [comment]
> I would also like to avoid backwards incompatible changes. IIUC pam_access is older than me and it had no hostname support for almost that entire time. The hostname feature was a backwards incompatible change, that surreptitiously transformed the meaning of existing configurations into allowing remote hosts. It did so without a mention in the NEWS file, and was then widely backported in distros like RHEL, thinking it a bugfix, I assume, which indicates to me that downstream did not understand this change. In my experience, it is vanishingly rare we can trust getaddrinfo to be accurate and tamper-proof, so I would not have opted-in to this change, given the choice, but until yesterday there was not a way to disable it.

I cannot speak for RedHat, but I would be surprised if the reason isn't similar for them: the feature got implemented because of high customer demand, nobody wants to maintain long lists of IPv4 and IPv6 addresses in a datacenter.
And many big networks are using this feature. By looking into our support files, if pam_access is used, I see rules depending on hosts (IP addresses, DNS names), I don't see rules for ttys or PAM services.
The support for TTYs is coming from the time when you had a terminal server and wanted to restrict access to the terminals, what is the use case today?
Same for PAM services, the history is to prevent users from installing cron jobs. This was important for the accounting at that time. But nobody is accounting user login times anymore today, or who is still using `acct`? I bet most people never heard of that tool. Beside that some distros switched already completely to systemd.timers.

So from what I see from data centers, removing support for TTYs or PAM services will not impact anybody, but removing DNS support will impact a huge number of people.

**Comment by pemensik** [comment]
Prefix syntax would be definitely a better idea. First it supports also IPv6, second it is somehow easy to verify the format. Minimal change would be to documentation, mentioning that any name may result in name resolution lookup (Not necessarily only DNS).

Okay, maybe keep all names without special prefix considered like hostnames and require prefix for DISPLAY, TTY or PAM service names. If only getaddrinfo numeric format were accepted, it would be okay. But when dns translation is done to each word used, it should allow syntax to avoid that. Allow bare tty names for backward compatibility, but warn about using bare tty or pam services in documentation and declare them deprecated. Maybe choose some prefix character to avoid using in dns, which would be removed from the name when testing it for other things. Could be one of ``[~^%$:]`` for example.

**Comment by ldv-alt** [comment]
> I would also like to avoid backwards incompatible changes. IIUC pam_access is older than me and it had no hostname support for almost that entire time. The hostname feature was a backwards incompatible change, that surreptitiously transformed the meaning of existing configurations into allowing remote hosts. It did so without a mention in the NEWS file, and was then widely backported in distros like RHEL,

I don't see 23393bef92c1e768eda329813d7af55481c6ca9f backported to fedora or centos.  I didn't check rhel, but I doubt it was backported there. @ikerexxe, please correct me if I'm wrong.

The first public release containing 23393bef92c1e768eda329813d7af55481c6ca9f was v1.5.3 released in May 2023, i.e. about 1.5 years ago.  Given that since 1997 and up to May 2023 pam_access didn't support that hostname resolving syntax, and it wasn't properly announced in the NEWS file, I suppose this new problematic hostname resolving syntax is not widely used yet, and we still can fix the syntax without risk of causing too much trouble to the users.

**Comment by ikerexxe** [comment]
> I don't see [23393be](https://github.com/linux-pam/linux-pam/commit/23393bef92c1e768eda329813d7af55481c6ca9f) backported to fedora or centos. I didn't check rhel, but I doubt it was backported there. @ikerexxe, please correct me if I'm wrong.

Yes, as Thorsten already mentioned, we at Red Hat also have enterprise customers which have a proper configuration and asked for this feature, thus it landed in recent RHEL8 and 9 versions. This feature is crucial for our enterprise customers who require the highest security standards to secure their systems and comply with current regulations.

Needless to say, the fixes that are being made in recent times are also being ported to these versions to prevent users from having any problems using this functionality.

**Comment by Fmiard** [comment]
After I made this fix commit as a patch into pam, I added a line to /etc/security/access.conf to configure:
![image](https://github.com/user-attachments/assets/3be273be-020a-419a-8d8a-89fdafbb05d3)
then added this line to the ssh configuration and added the debug parameter：
![image](https://github.com/user-attachments/assets/e2f1bd7b-9904-41b0-9053-bd3f93d8cd54)
When I log in using ssh I should see this line in /var/log/secure:
"network_netmask_match: tok=tty1 is a TTY"
unfortunately I don't. However, when I logged in with ssh after disabling selinux with “setenforce 0”, this line appeared in the /var/log/secure logs!
I would like to ask if this is a reasonable situation, and if turning selinux on or off will affect the fix?

**Comment by thkukuk** [comment]
> I would like to ask if this is a reasonable situation, and if turning selinux on or off will affect the fix?

If your SELinux policy forbids your application to access `/dev`, then yes, this situation is to be expected.
Turning off SELinux is never a valid fix or workaround. The correct fix is to adjust your SELinux policy.

**Comment by Fmiard** [comment]
> > I would like to ask if this is a reasonable situation, and if turning selinux on or off will affect the fix?
> 
> If your SELinux policy forbids your application to access `/dev`, then yes, this situation is to be expected. Turning off SELinux is never a valid fix or workaround. The correct fix is to adjust your SELinux policy.

Thank you very much for your answer！I also have some questions.
 1)When ssh logging in using the above access.conf configuration method and using the default SELinux policy which is not configured to allow applications to access /dev, there is no log “network_netmask_match: tok=tty1 is a TTY “. Does this mean that there is something wrong with the execution of the “is_device” function that it doesn't go to the if conditional statement to perform the log print?
2)Also, I noticed that when there is no “network_netmask_match: tok=tty1 is a TTY” log print, there is this line in the /var/log/secure “cannot resolve hostname tty1”, this is the program into the ‘getaddrinfo’ function of the if statement will be executed to print the log information. Does this mean that pam_access can't correctly determine what “tty1” is when the user is not actively configuring the SELinux policy?
![image](https://github.com/user-attachments/assets/c03eb39a-323d-42ef-b6ee-d9cbecf73b15)

**Comment by thkukuk** [comment]
> Thank you very much for your answer！I also have some questions. 1)When ssh logging in using the above access.conf configuration method and using the default SELinux policy which is not configured to allow applications to access /dev, there is no log “network_netmask_match: tok=tty1 is a TTY “. Does this mean that there is something wrong with the execution of the “is_device” function that it doesn't go to the if conditional statement to perform the log print? 

Of course is the execution of the "is_device" function not working if you forbid access to /dev. Quite logical, or?

> 2)Also, I noticed that when there is no “network_netmask_match: tok=tty1 is a TTY” log print, there is this line in the /var/log/secure “cannot resolve hostname tty1”,

Correct, see 1).

> this is the program into the ‘getaddrinfo’ function of the if statement will be executed to print the log information. Does this mean that pam_access can't correctly determine what “tty1” is when the user is not actively configuring the SELinux policy? 

If your SELinux forbids your application to access /dev, of course can pam_access not determine what "tty1" is, how should it do that?

**Comment by aditjha-msft** [comment]
Hi @ldv-alt and @thkukuk , I am a developer at Microsoft for Azure Linux, and was curious about some of the details regarding: https://nvd.nist.gov/vuln/detail/CVE-2024-10963

Because of the wording on some of the previous replies, I am a bit confused whether #447 technically _fixes_ CVE-2024-10963 or is it instead the PR that _introduced_ the vulnerability described in CVE-2024-10963.  

We have images with 1.5.1 and 1.5.3 versions of pam so we are trying to act accordingly. 

Apologies in advance for the clarification requested, I appreciate it.

**Comment by aditjha-msft** [comment]
@rpigott thank you very much for your response. For 1.5.3, is #854 enough to consider a fix for this CVE? We are deciding whether to act on this CVE now or wait until a future patch. appreciate the help!

**Comment by rpigott** [comment]
#854 introduces an option, "nodns", which will revert to the old behavior if you use it. If you use that option in your pam configuration, you can use tty/service names in access.conf without being affected by the CVE.

**Comment by aditjha-msft** [comment]
Is there a plan to have a separate fix for the CVE that would not require a manual option change. Since we ship Azure Linux without a custom pam configuration, we'd strongly prefer to handle this CVE by relying on default behavior. Don't mind waiting on an upstream fix, but want to check in case #854 is going to be the only upstream PR related to this CVE. Thanks in advance!

**Comment by liuzheng** [comment]
Upgraded still stuck at **cron** **crond** 
```
# access.conf
+:root:cron crond :0 tty1 tty2 tty3 tty4 tty5 tty6 
```

![Image](https://github.com/user-attachments/assets/76483163-a91e-4f62-a374-4348ca9bacd9)

![Image](https://github.com/user-attachments/assets/ebdb3e46-1aa0-4ad3-8be9-ffc311c0c5ed)

**Comment by simorex80** [comment]
Any ideas when the vulnerability CVE-2024-10963 will be fixed?
Thanks

**Comment by hartmans** [comment]
> > I don't see [23393be](https://github.com/linux-pam/linux-pam/commit/23393bef92c1e768eda329813d7af55481c6ca9f) backported to fedora or centos. I didn't check rhel, but I doubt it was backported there. [@ikerexxe](https://github.com/ikerexxe), please correct me if I'm wrong.
> 
> Yes, as Thorsten already mentioned, we at Red Hat also have enterprise customers which have a proper configuration and asked for this feature, thus it landed in recent RHEL8 and 9 versions. This feature is crucial for our enterprise customers who require the highest security standards to secure their systems and comply with current regulations.
> 
> Needless to say, the fixes that are being made in recent times are also being ported to these versions to prevent users from having any problems using this functionality.

Hi.  I'm the Debian PAM maintainer. I also happen to be reasonably familiar with Internet security.ying to understand the requirements for the pam_access hostname feature to be part of a high security solution, even in a datacenter.
It seems like you would need to be able to trust the results of getaddrinfo and to trust that IP spoofing is not happening.
I can believe that a lot of data center networks have features that would limit the effectiveness of IP spoofing.
I don't see a lot of use of DNsSec in datacenters though, so what is giving you confidence in the getaddrinfo results?
Are you assuming local deployment of authenticated DNS over TLS or DNS over HTTP?

Or is this more of a compliance thing than a security thing? Someone has an audit control that requires they limit access to a particular set of hosts, and this gives them a plausible story for meeting that control, even if the answers from getaddrinfo is not fully authenticated.

Or am I missing some common deployment that gives you confidence in getaddrinfo results some other way?

I think what I'm trying to say is that I can easily believe  enterprise customers are requesting this, and I can easily believe there is a required audit control this helps people satisfy.
But I'm wondering whether it is actually a secure solution to that control, and if so what assumptions in real-world data centers make that assumption true.

**Comment by rpigott** [comment]
FWIW my security concern was not that getaddrinfo could never be made reliable on specific networks for specific hosts with specific requirements and controls, resolving specific domains, but rather that this feature, as implemented, introduced such a requirement for every user of pam_access, retroactively changing the meaning of existing access.conf configurations that were otherwise working as intended for decades, without so much as an offhand mention in the changelog.

Surely it would have been possible to introduce this feature with new syntax, for the new users that wanted to use it going forward with the new version of pam, but we don't live in that timeline. For all the reasons I outlined above, I would have preferred for this change to be reverted and the feature reintroduced in some other way.

> I don't see a lot of use of DNsSec in datacenters though, so what is giving you confidence in the getaddrinfo results?

I have personally invested plenty of time and code into improving the dnssec implementation in sd-resolved, and I will invest more time to have the _completely_ insecure LLMNR resolver disabled by default, at least in part because it does not match the bold assumptions apparently made by developers around the security of interfaces like getaddrinfo.

But for clarity, even absent alternative name databases, DNSSEC is not sufficient to ensure the security of getaddrinfo for this purpose. DNSSEC is capable of rejecting bogus responses for signed domains, but generally will permit unauthenticated responses for unsigned domains for compatibility with the vast majority of unsigned domains that populate the internet. With the libc resolver, this behavior is indicated only by the AD bit in the response with trust-ad configured, and not conferred at all in the addrinfo structs returned by getaddrinfo. Software like openssh that relies on the authenticity of dns records, e.g. SSHFP record types, generally uses the res_query interface to get this additional info from the system resolver instead of getaddrinfo. Otherwise, if there is at least one unsigned search domain configured, DNSSEC will not protect the user from this feature.
